infoschema/cache.go

```diff
@@ -98,48 +139,70 @@ func (h *InfoCache) GetBySnapshotTS(snapshotTS uint64) InfoSchema {
 	defer h.mu.RUnlock()
 
 	infoschema_metrics.GetTSCounter.Inc()
-	if snapshotTS >= h.maxUpdatedSnapshotTS {
-		if len(h.cache) > 0 {
-			infoschema_metrics.HitTSCounter.Inc()
-			return h.cache[0]
-		}
+	if schema, err := h.getSchemaByTimestampNoLock(snapshotTS); err == nil {
+		infoschema_metrics.HitTSCounter.Inc()
+		return schema
 	}
 	return nil
 }
 
 // Insert will **TRY** to insert the infoschema into the cache.
 // It only promised to cache the newest infoschema.
 // It returns 'true' if it is cached, 'false' otherwise.
-func (h *InfoCache) Insert(is InfoSchema, snapshotTS uint64) bool {
+// snapshotTS is the timestap of the schema update, schemaTS is the timestamp of the schema taking effect
+func (h *InfoCache) Insert(is InfoSchema, snapshotTS, schemaTS uint64) bool {
 	h.mu.Lock()
 	defer h.mu.Unlock()
 
 	version := is.SchemaMetaVersion()
+
+	// assume this is the timestamp order as well
 	i := sort.Search(len(h.cache), func(i int) bool {
-		return h.cache[i].SchemaMetaVersion() <= version
+		return h.cache[i].infoschema.SchemaMetaVersion() <= version
 	})
 
+	// keep maxUpdatedSnapshotTS for backward compatibility
 	if h.maxUpdatedSnapshotTS < snapshotTS {
```

Comment:
We don't need this, if we really give accurate timestamp. Your solution will be an improvement over `maxUpdatedSnapshotTS`.